### PR TITLE
Add prometheus scraping

### PR DESCRIPTION
**Pull request by sjug**
Add CPU and memory scraping directly from Prometheus, including support behind insecure TLS authenticated proxy as is the case with OpenShift.

### CODE REVIEW

File: cmd/scraper/main.go [hunk 1]
```diff
@@ -5,40 +5,54 @@ import (
...
-
-func initFlags() {
-	flag.StringVar(&searchDir, "i", "/var/lib/pbench-agent/benchmark_result/tools-default/", "pbench run result directory to parse")
-	flag.StringVar(&resultDir, "o", "/tmp/", "output directory for parsed CSV result data")
-	flag.StringVar(&processString, "proc", "openshift_start_master_api_,openshift_start_master_controll,hyperkube_kubelet_,openshift_start_node_,etcd,dockerd-current_,elasticsearc,prometheus_,systemd_--switched-root,openshift_start_network_,ovs-vswitchd_unix,openshift-router,fluentd,kibana,heapster,crio", "list of processes to gather")
-	flag.StringVar(&blockString, "blkdev", "sda-write,sda-read,vda-write,vda-read,xvda-write,xvda-read,xvdb-write,xvdb-read,nvme0n1-write,nvme0n1-read", "List of block devices")
-	flag.StringVar(&netString, "netdev", "eth0-rx,eth0-tx", "List of network devices")
+func initFlags() (cfg config.ScrapeConfig) {
+	flag.BoolVar(&cfg.EnablePbenchFlag, "pbench", false, "scrape pbench results")
+	flag.BoolVar(&cfg.EnablePromtheusFlag, "prometheus", false, "scrape prometheus endpoint")
```
Comment (jmencak) [inline cmd/scraper/main.go]: `EnablePrometheusFlag`?  Similar replacement elsewhere.

File: pkg/prometheus/prometheus.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+		fmt.Printf("Error creating Prometheus client: %v\n", err)
+		c.Errors = append(c.Errors, err)
+	} else {
+		c.API = v1.NewAPI(client)
+	}
+}
+
+// DoPrometheusQuery will run queries against Prometheus endpoint
+func DoPrometheusQuery(cfg config.ScrapeConfig) {
+	//cfg.UrlFlag, cfg.TokenFlag, cfg.StepFlag, cfg.DurationFlag, cfg.InsecureTLSFlag
```
Comment (jmencak) [inline pkg/prometheus/prometheus.go]: `s|//cfg|// cfg|`

File: pkg/prometheus/prometheus.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	// url, token, stepFlag string, durationFlag int, insecureTLS bool
+	config := newPrometheusConfig(cfg.UrlFlag, cfg.TokenFlag, cfg.InsecureTLSFlag)
+	config.newPrometheusAPI()
+	if len(config.Errors) != 0 {
+		fmt.Printf("Client Error %v\n", config.Errors[0])
+	}
+
+	queries := map[string]string{
+		"cpu":    "sum(rate(container_cpu_usage_seconds_total{job=\"kubelet\", image!=\"\", container!=\"POD\"}[5m])) by (namespace)",
```
Comment (jmencak) [inline pkg/prometheus/prometheus.go]: Would raw string literals work and be easier to read?  I.e. 
```
`sum(rate(container_cpu_usage_seconds_total{job="kubelet", image!="", container!="POD"}[5m])) by (namespace)`
```

File: pkg/prometheus/prometheus.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		"memory": "sum(container_memory_usage_bytes{container_name!=\"\"}) by (namespace)",
```
Comment (jmencak) [inline pkg/prometheus/prometheus.go]: Raw string literals?

File: pkg/prometheus/prometheus.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	if err != nil {
+		fmt.Printf("Error parsing step duration %s\n", cfg.StepFlag)
+	}
+	r := v1.Range{Start: start, End: end, Step: step}
+
+	var results []result.ResultType
+	for resource, query := range queries {
+		queryResult, _, err := config.API.QueryRange(context.Background(), query, r)
+		if err != nil {
+			fmt.Printf("Query Error: %v\n", err)
```
Comment (jmencak) [inline pkg/prometheus/prometheus.go]: You may want to say "Prometheus query error" or something like that.  "Query Error" seems too generic to me.  Similarly below.  Any reason for uppercasing Error?

File: pkg/prometheus/prometheus.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+			var newData rawData
+			for _, v := range j.Metric {
+				newData.Name = string(v)
+			}
+			for _, v := range j.Values {
+				newData.Values = append(newData.Values, float64(v.Value))
+			}
+			series.Tags = append(series.Tags, newData)
+		}
+		fmt.Printf("Series: %+v\n", series)
```
Comment (jmencak) [inline pkg/prometheus/prometheus.go]: Is this for debugging or part of output?

**Review by jmencak — COMMENTED** [review]
Only a few initial comments, as I guess this is still WIP.  Couldn't compile due to missing packages, go.mod is probably the best way forward for now as dissussed.

File: pkg/prometheus/prometheus.go (hunk 5 above)
Comment (sjug) [reply pkg/prometheus/prometheus.go]: Fixed.

File: cmd/scraper/main.go (hunk 1 above)
Comment (sjug) [reply cmd/scraper/main.go]: Fixed.

File: pkg/prometheus/prometheus.go (hunk 2 above)
Comment (sjug) [reply pkg/prometheus/prometheus.go]: Removed

File: pkg/prometheus/prometheus.go (hunk 6 above)
Comment (sjug) [reply pkg/prometheus/prometheus.go]: It's for debugging

File: pkg/prometheus/prometheus.go (hunk 3 above)
Comment (sjug) [reply pkg/prometheus/prometheus.go]: Good idea

File: pkg/prometheus/prometheus.go (hunk 4 above)
Comment (sjug) [reply pkg/prometheus/prometheus.go]: Changed

**Comment by sjug** [comment]
@jmencak Thanks for the quick review, I've added the `go.mod` and you should be good to go.

File: pkg/prometheus/prometheus.go (hunk 6 above)
Comment (jmencak) [reply pkg/prometheus/prometheus.go]: If it is for debugging, shouldn't debugging info go to stderr?

**Comment by jmencak** [comment]
Is there a reason to require golang 1.13?  Also, the purpose of makefiles is really to make things easy for users to build.  I'd expect some env var setting to use go modules (GO111MODULE=on) and pull the dependencies if you decided not to vendor them in or at least document how to build.

**Comment by sjug** [comment]
Hey @jmencak, that's just what I use on my system so that's what the `go.mod` was populated with. It's nice as you don't need that env var, as long as you're working outside of the $GOPATH it uses modules by default now. 

For me to build on 1.13 I just clone outside of the $GOPATH, run `make` and that's it.

File: pkg/prometheus/prometheus.go (hunk 6 above)
Comment (sjug) [reply pkg/prometheus/prometheus.go]: Fixed.

**Pull request merged**